File: src/clustr/tools/write/container_snapshots.py
"""
Write tools for LXC container snapshot management.

Annotations:
  - create_container_snapshot:   destructiveHint = False
  - delete_container_snapshot:   destructiveHint = True
  - rollback_container_snapshot: destructiveHint = True
"""

from __future__ import annotations

import logging
import re
from typing import Annotated, Any, cast

from mcp.server.fastmcp import FastMCP
from mcp.types import ToolAnnotations
from pydantic import Field

from clustr.proxmox.client import get_client, proxmox_post
from clustr.tools import safe

logger = logging.getLogger(__name__)

_SNAP_NAME_RE = re.compile(r"^[a-zA-Z0-9_\-]{1,40}$")

_SAFE_WRITE = ToolAnnotations(readOnlyHint=False, destructiveHint=False)
_DESTRUCTIVE = ToolAnnotations(readOnlyHint=False, destructiveHint=True)

_Node = Annotated[str, Field(description="Node name")]
_CtId = Annotated[int, Field(ge=100, description="Container ID")]


# ---------------------------------------------------------------------------
# Tool implementations
# ---------------------------------------------------------------------------


def _create_container_snapshot(
    node: str, ctid: int, snapname: str, description: str = ""
) -> str:
    params: dict[str, Any] = {"snapname": snapname}
    if description:
        params["description"] = description
    return cast(
        str,
        proxmox_post(
            lambda: get_client().nodes(node).lxc(ctid).snapshot.post(**params)
        ),
    )


def _delete_container_snapshot(node: str, ctid: int, snapname: str) -> str:
    return cast(
        str,
        proxmox_post(
            lambda: get_client().nodes(node).lxc(ctid).snapshot(snapname).delete()
        ),
    )


def _rollback_container_snapshot(node: str, ctid: int, snapname: str) -> str:
    return cast(
        str,
        proxmox_post(
            lambda: get_client()
            .nodes(node)
            .lxc(ctid)
            .snapshot(snapname)
            .rollback.post()
        ),
    )


# ---------------------------------------------------------------------------
# Tool registration
# ---------------------------------------------------------------------------


def register(mcp: FastMCP) -> None:
    """Register all container snapshot tools onto the given FastMCP instance."""

    @mcp.tool(
        name="create_container_snapshot",
        title="Create Container Snapshot",
        description=(
            "Create a snapshot of an LXC container. "
            "Snapshot names must be alphanumeric with hyphens/underscores, "
            "max 40 chars."
        ),
        annotations=_SAFE_WRITE,
    )
    async def create_container_snapshot(
        node: _Node,
        ctid: _CtId,
        snapname: Annotated[
            str,
            Field(
                description=(
                    "Snapshot name (alphanumeric, hyphens, underscores; "
                    "max 40 chars)"
                )
            ),
        ],
        description: Annotated[str, Field(description="Optional description")] = "",
    ) -> str:
        def _do() -> str:
            if not _SNAP_NAME_RE.match(snapname):
                return (
                    "Error: Snapshot name must be alphanumeric with "
                    "hyphens/underscores only, max 40 characters."
                )
            task_id = _create_container_snapshot(node, ctid, snapname, description)
            return (
                f"✅ Snapshot **{snapname}** creation started for "
                f"container {ctid} on {node}.\n"
                f"Task ID: `{task_id}`\n\n"
                f"Use `list_container_snapshots` to confirm when complete."
            )

        return await safe("create_container_snapshot", _do)

    @mcp.tool(
        name="delete_container_snapshot",
        title="Delete Container Snapshot",
        description=(
            "Permanently delete a snapshot of an LXC container. "
            "This action cannot be undone."
        ),
        annotations=_DESTRUCTIVE,
    )
    async def delete_container_snapshot(
        node: _Node,
        ctid: _CtId,
        snapname: Annotated[str, Field(description="Exact snapshot name to delete")],
    ) -> str:
        def _do() -> str:
            task_id = _delete_container_snapshot(node, ctid, snapname)
            return (
                f"✅ Snapshot **{snapname}** deletion started for "
                f"container {ctid} on {node}.\n"
                f"Task ID: `{task_id}`"
            )

        return await safe("delete_container_snapshot", _do)

    @mcp.tool(
        name="rollback_container_snapshot",
        title="Rollback Container to Snapshot",
        description=(
            "Rollback an LXC container to a previous snapshot state. "
            "WARNING: All changes after the snapshot was taken will be lost. "
            "The container must be stopped before rolling back."
        ),
        annotations=_DESTRUCTIVE,
    )
    async def rollback_container_snapshot(
        node: _Node,
        ctid: _CtId,
        snapname: Annotated[str, Field(description="Snapshot name to roll back to")],
    ) -> str:
        def _do() -> str:
            task_id = _rollback_container_snapshot(node, ctid, snapname)
            return (
                f"✅ Rollback to snapshot **{snapname}** started for "
                f"container {ctid} on {node}.\n"
                f"Task ID: `{task_id}`\n\n"
                f"⚠️ All changes after this snapshot was taken have been discarded."
            )

        return await safe("rollback_container_snapshot", _do)

File: clustr-ts/src/tools/write/containerSnapshots.ts
/**
 * Write tools for LXC container snapshots. LXC snapshots have no RAM state.
 */

import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";

import { proxmoxDelete, proxmoxPost } from "../../proxmox.js";
import { needsConfirm, safe } from "../../safe.js";

const SAFE_WRITE = { readOnlyHint: false, destructiveHint: false } as const;
const DESTRUCTIVE = { readOnlyHint: false, destructiveHint: true } as const;
const SNAP_NAME = /^[a-zA-Z0-9_-]{1,40}$/;

const node = z.string().describe("Node name");
const ctid = z.number().int().min(100).describe("Container ID");
const confirm = z
  .boolean()
  .default(false)
  .describe(
    "Must be true to execute this destructive operation. When false (default), " +
      "returns a confirmation prompt without acting.",
  );

export function register(server: McpServer): void {
  server.registerTool(
    "create_container_snapshot",
    {
      title: "Create Container Snapshot",
      description:
        "Create a snapshot of an LXC container. Names: alphanumeric/hyphens/" +
        "underscores, max 40 chars.",
      inputSchema: {
        node,
        ctid,
        snapname: z.string().describe("Snapshot name (alphanumeric, -, _; max 40)"),
        description: z.string().optional().describe("Optional description"),
      },
      annotations: SAFE_WRITE,
    },
    async ({ node: n, ctid: id, snapname, description }) =>
      safe("create_container_snapshot", async () => {
        if (!SNAP_NAME.test(snapname)) {
          return "Error: Snapshot name must be alphanumeric with hyphens/underscores only, max 40 characters.";
        }
        const body: Record<string, string> = { snapname };
        if (description) body.description = description;
        const task = await proxmoxPost(`/nodes/${n}/lxc/${id}/snapshot`, body);
        return (
          `✅ Snapshot **${snapname}** creation started for container ${id} on ${n}.\n` +
          `Task ID: \`${String(task)}\`\n\nUse \`list_container_snapshots\` to confirm.`
        );
      }),
  );

  server.registerTool(
    "delete_container_snapshot",
    {
      title: "Delete Container Snapshot",
      description:
        "Permanently delete a snapshot of an LXC container. Cannot be undone.",
      inputSchema: {
        node,
        ctid,
        snapname: z.string().describe("Exact snapshot name to delete"),
        confirm,
      },
      annotations: DESTRUCTIVE,
    },
    async ({ node: n, ctid: id, snapname, confirm: ok }) =>
      safe("delete_container_snapshot", async () => {
        if (!SNAP_NAME.test(snapname)) {
          return "Error: invalid snapshot name (alphanumeric, hyphens/underscores, max 40).";
        }
        if (!ok) {
          return needsConfirm(
            "delete snapshot",
            `**${snapname}** of container ${id} on ${n}`,
          );
        }
        const task = await proxmoxDelete(
          `/nodes/${n}/lxc/${id}/snapshot/${snapname}`,
        );
        return `✅ Snapshot **${snapname}** deletion started for container ${id} on ${n}.\nTask ID: \`${String(task)}\``;
      }),
  );

  server.registerTool(
    "rollback_container_snapshot",
    {
      title: "Rollback Container to Snapshot",
      description:
        "Roll an LXC container back to a snapshot. WARNING: all changes after " +
        "the snapshot are lost. The container should be stopped first.",
      inputSchema: {
        node,
        ctid,
        snapname: z.string().describe("Snapshot name to roll back to"),
        confirm,
      },
      annotations: DESTRUCTIVE,
    },
    async ({ node: n, ctid: id, snapname, confirm: ok }) =>
      safe("rollback_container_snapshot", async () => {
        if (!SNAP_NAME.test(snapname)) {
          return "Error: invalid snapshot name (alphanumeric, hyphens/underscores, max 40).";
        }
        if (!ok) {
          return needsConfirm(
            "roll back to snapshot",
            `**${snapname}** on container ${id} (${n}) — discarding all later changes`,
          );
        }
        const task = await proxmoxPost(
          `/nodes/${n}/lxc/${id}/snapshot/${snapname}/rollback`,
        );
        return (
          `✅ Rollback to snapshot **${snapname}** started for container ${id} on ${n}.\n` +
          `Task ID: \`${String(task)}\`\n\n⚠️ All changes after this snapshot have been discarded.`
        );
      }),
  );
}
